### DISCUSSION TITLE
Add region markers  into exception data

### DISCUSSION THREAD

**Discussion by nohwnd** [fluentassertions/fluentassertions#1826] [Ideas]
When expected or actual object is big it is very hard to find the assertion message. Could additional data that could be used for formatting be added to the exception Data dictionary?

For example (newlines are mine):

```
FluentAssertions.Execution.AssertionFailedException: Did not expect 
fixture.TestRunEventsRegistrar.LoggedWarnings 
{"Test run detected DLL(s) which were built for different framework and platform versions.
Following DLL(s) do not match current settings, which are .NETCoreApp,Version=v1.0 framework and X64 platform.
<...10 more lines of output>"} to contain a match of "Test run detected DLL(s) which were built for different framework*".
```

Is really hard for me to read. It would be much easier if the assertion message was printed in a different color than the actual and expected value.

Or if I could have the different colors and print it as: 
```
FluentAssertions.Execution.AssertionFailedException: Did not expect fixture.TestRunEventsRegistrar.LoggedWarnings 
to contain a match of "Test run detected DLL(s) which were built for different framework*".

But actual value was: 
{"Test run detected DLL(s) which were built for different framework and platform versions.
Following DLL(s) do not match current settings, which are .NETCoreApp,Version=v1.0 framework and X64 platform.
<...10 more lines of output>"} 
```

I think both could be achieved by adding a dictionary<long, int> of markers to the Exception.Data[FormattingMarkers], that would say (example values):

0, 1
120, 2
746, 1
948, 3

Where key is the index in the string, and value is an enum defined in FluentAssertions, but used as a number, meaning 1=message, 2=actual, 3=expected.

The consumer of that exception could then split the string and format it as it sees fit, with different coloring (also using ansi color codes), or by adding newlines between the message and object, or re-organizing the message altogether. It would also have no impact on the current consumers, other than a tiny perf impact on creating the dictionary for a failing test.

**Comment by nohwnd** [comment]
This would be similar to semantic tokens in LSP, which are a bit more complex as there are many different types of tokens, and lines and whitespace are involved, because only part of the document might need to update.

Lsp first sends a legend that contains the map from token type to the number (for fluent assertions that would be 1= mesage, 2 = actual, 3 = expected). And then for every request sends back a map of locations to the token type. 

https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide#semantic-token-provider

For fluent assertion there are <5 token types, that will probably rarely expand, and will never change order. So adding an additional legend to every exception seems unnecessary. 

And whitespace and lines can be safely ignored because we always get the whole string.

**Comment by dennisdoomen** [comment]
Just played with the ANSI escape codes a bit

<img width="444" alt="image" src="https://user-images.githubusercontent.com/572734/155184567-86673d61-d513-4db3-b773-542ff7d9a38d.png">

**Reply by nohwnd** [reply]
Works for me, if you really want to commit to choosing a color and then have people complain that in VS with a custom theme it renders with a slightly wrong contrast. :)
